src/background/utils/requests.js

```diff
@@ -188,58 +193,37 @@ function decodeBody(obj) {
 // });
 
 // Modifications on headers
-{
-  function onBeforeSendHeaders(details) {
-    const headers = details.requestHeaders;
-    let newHeaders = [];
-    const vmHeaders = {};
-    headers.forEach((header) => {
-      // if (header.name === 'VM-Task') {
-      //   tasks[details.requestId] = header.value;
-      // } else
-      if (header.name.startsWith('VM-')) {
-        vmHeaders[header.name.slice(3)] = header.value;
-      } else {
-        newHeaders.push(header);
-      }
-    });
-    const reqId = vmHeaders.Verify;
-    if (reqId) {
-      const req = requests[reqId];
-      if (req) {
-        delete vmHeaders.Verify;
-        verify[details.requestId] = reqId;
-        req.coreId = details.requestId;
-        Object.keys(vmHeaders).forEach((name) => {
-          if (isSpecialHeader(name.toLowerCase())) {
-            newHeaders.push({ name, value: vmHeaders[name] });
-          }
-        });
-        if (req.anonymous) {
-          // Drop cookie in anonymous mode
-          newHeaders = newHeaders.filter(({ name }) => name.toLowerCase() !== 'cookie');
-        }
-      }
-    }
-    return { requestHeaders: newHeaders };
-  }
-  const filter = {
-    urls: ['<all_urls>'],
-    types: ['xmlhttprequest'],
+const customHeaders = Object.create(null);
+function onBeforeSendHeaders({ requestHeaders: headers, requestId }) {
+  // only the first call during a redirect/auth chain will have VM-Verify header
+  const reqId = (headers.find(h => h.name === VM_VERIFY) || {}).value || verify[requestId];
+  const req = reqId && requests[reqId];
+  if (!reqId || !req) return { requestHeaders: headers };
+  verify[requestId] = reqId;
+  req.coreId = requestId;
+  return {
+    requestHeaders: headers
+    .filter(({ name }) => name !== VM_VERIFY && (!req.anonymous || !/^cookie$/i.test(name)))
```

Comment:
I wonder if `filter` should come after `concat` so that a `cookie` is removed from custom headers as well? OTOH it's the userscript's responsibility to provide sensible noncontradictory params...

Reply:
Nevermind, previously they were filtered so I'll just do it here as well.